internal/util/http.go

```diff
@@ -0,0 +1,174 @@
+	// Since we create this client once during startup,
+	// we want each connection to be fresh, hence the non-default transport
+	// See https://github.com/golang/go/issues/43905 and https://github.com/golang/go/issues/23427
+	// for discussion of problems with long-lived connections
+	client = &http.Client{
+		Timeout: 5 * time.Second,
+		Transport: &http.Transport{
+			DisableKeepAlives:   true,
+			MaxIdleConnsPerHost: -1,
+		},
+	}
+}
+
+// Extensible struct for client update data
+type clientUpdateData struct {
+	Version string `json:"version"`
+}
+
+func addHeader(req *http.Request, versionString string) (*http.Request, error) {
+	updateInfo := &clientUpdateData{
+		Version: versionString,
+	}
+	marshaled, err := json.Marshal(updateInfo)
+	if err == nil {
+		updateInfoString := string(marshaled)
+		req.Header.Add(clientHeader, updateInfoString)
+	}
+
+	return req, err
+}
+
+func checkForUpdates(versionString string, backoff wait.Backoff) (string, error) {
+	var res *http.Response
+	var err error
+	var bodyBytes []byte
+
+	// Prep request
+	req, err := http.NewRequest("GET",
+		updateCheckURL,
+		nil)
+	if err == nil {
+		req, err = addHeader(req, versionString)
+	}
+
+	// wait.ExponentialBackoff will retry the func according to the backoff object until
+	// (a) func returns done as true or
+	// (b) the backoff settings are exhausted,
+	// i.e., the process hits the cap for time or the number of steps
+	// The anonymous function here sets certain preexisting variables (res, err)
+	// which are then used by the surrounding `checkForUpdates` function as part of the return
+	if err == nil {
+		_ = wait.ExponentialBackoff(
+			backoff,
+			func() (done bool, backoffErr error) {
+				// We can't close the body of this response in this block since we use it outside
+				// so ignore this linting error
+				res, err = client.Do(req) //nolint:bodyclose
+				// This is a very basic check, ignoring nuances around
+				// certain StatusCodes that should either prevent or impact retries
+				if err == nil && res.StatusCode == http.StatusOK {
+					return true, nil
+				}
+				if err == nil {
+					err = fmt.Errorf("received StatusCode %d", res.StatusCode)
+				}
+				// Return false, nil to continue checking
+				return false, nil
+			})
+	}
+	// If the final value of err is nil and the final res.StatusCode is OK,
+	// we can go on with reading the body of the response
+	if err == nil && res.StatusCode == http.StatusOK {
+		defer res.Body.Close()
+		bodyBytes, err = ioutil.ReadAll(res.Body)
+	}
+
+	// TODO: Parse response and log info for user on potential updates
+	if err == nil {
+		return string(bodyBytes), nil
+	}
+
+	return "", err
+}
+
+// CheckForUpdatesScheduler invokes the check func when the operator starts
+// and then on the given period schedule
+func CheckForUpdatesScheduler(versionString string, channel chan bool) {
+	ctx := context.Background()
+	log := logging.FromContext(ctx)
+	defer func() {
+		if err := recover(); err != nil {
```

Comment:
❓ I imagine we want to continue trying to check for updates when something panics. Should there be a similar defer+recover in `checkForUpdates`?

🤔 Recovering from panic in each HTTP call seems like too much. One recovery for all of `checkForUpdates` is probably enough. Perhaps there it can convert to an `error` and return.

Reply:
OK, makes a lot of sense, I added a defer+recover to `checkForUpdates`, so now a single panicking call shouldn't interfere with future calls.

The downside of this is that since I am recovering from panics in `checkForUpdates`, I'm not sure how to test panic recovery from `CheckForUpdatesScheduler` since I was originally hijacking my mocked client call inside `checkForUpdates` to cause a panic.